### PR TITLE
Allow generalized ClassifierTrainer inheritance, keeping backwards compatibility with the MiniBatchTrainer

### PR DESCRIPTION
**Pull request by gnieuwenhuis**
## Goal
To allow more control over the training loop in the mermaid classifier (see for example https://github.com/data-mermaid/mermaid-classifier/pull/18)

To accomplish this, we are allowing inheritance of the ClassifierTrainer here. This in turn allows defining custom training logic within an inherited trainer class. 

See this discussion for more background: https://github.com/data-mermaid/mermaid-classifier/issues/17

## These changes include
- make it possible to inherit from the ClassifierTrainer, and use that inherited trainer instead of the default MiniBatchTrainer
- If you do not specify a custom trainer, behaviour remains unchanged, and the default MiniBatchTrainer is still used.

### CODE REVIEW

**Comment by StephenChan** [comment]
Thanks @gnieuwenhuis - looks great all around.

- I like how you added not just the ClassifierTrainer instance option for `trainer`, but also the class path option, taking advantage of the fact that ClassifierTrainer doesn't take any init parameters (unlike FeatureExtractor). This enables a user to specify a `trainer` without having to import the trainer class first.

- Nice to see how support for the old `trainer_name` message parameter is retained as well. For the record, I actually have been less careful with changing the pyspacer message APIs, justifying a lack of deprecation cycles with pyspacer's "0.x" version numbering (something 1.0 or above would indicate a more stable package). But I think it's good to get into this habit of compatibility between consecutive pyspacer versions... the only reason it's not a problem for CoralNet is that we just accept some downtime during server updates, which lets us finish all submitted pyspacer jobs before updating the code. Anyway, this is good to have!

- The ClassifierTrainer changes themselves look good, and consistent with FeatureExtractor. Good idea on adding this check. I haven't actually thought of using `isinstance(<something>, type)` before.

  ```python
  if not (
      isinstance(trainer_class, type)
      and issubclass(trainer_class, ClassifierTrainer)
  )
  ```

- Defining the built-in trainers in `train_classifier.py` instead of `config.py`: also good IMO.

- The tests look complete and clean. Probably worth copying many of these test ideas into the feature extractor tests later.

I will run some tests myself, including with CoralNet, to check that everything's good to go. Otherwise the only to-do point on my mind is on `README.md`; the comment saying "'minibatch' is currently the only trainer that spacer defines." should be expanded to briefly explain that custom trainers are possible too.

**Comment by StephenChan** [comment]
Thanks for the README update. I also see a `CLAUDE.md` that you might've added accidentally though?

Testing on my end was all good.

One other comment I forgot to make: I was originally thinking of moving one or more of train(), evaluate_classifier(), and calc_acc() from train_utils.py into ClassifierTrainer or MiniBatchTrainer. However, at this point it's not entirely clear how that would increase ease of use, particularly when there is still only one built-in trainer class. So for now, I think it makes sense to not fix what's not broken.

**Comment by gnieuwenhuis** [comment]
Sorry about the added file - it was mistakenly committed and I'll remove right away. I've been using AI to help explore, ideate, and suggest solutions lately, which is why it was there to begin with.

I see your idea with the train(), evaluate_classifier() and calc_acc(), and I thought about it a little just now. For what its worth, I think that the evaluate_classifier() and calc_acc() are both functions that act more like building blocks which are easily useable in different ways if they remain in train_utils.py. The train() method is core functionality to the MiniBatchTrainer, and I think it would improve class cohesion to put it with that trainer. But I wouldn't change the abstract ClassifierTrainer to avoid overly constraining custom trainers (my thought is a little selfish, because I'm thinking about how I'm defining the mermaid-classifier custom trainer, and prefer the freedom that is available right now).

I'm happy making some modifications here if you want to to do that - my guess is that if it doesn't get done now it may not get done in the future either.

**Comment by StephenChan** [comment]
> Sorry about the added file - it was mistakenly committed and I'll remove right away. I've been using AI to help explore, ideate, and suggest solutions lately, which is why it was there to begin with.

No worries!

> ... The train() method is core functionality to the MiniBatchTrainer, and I think it would improve class cohesion to put it with that trainer. ... I'm happy making some modifications here if you want to to do that - my guess is that if it doesn't get done now it may not get done in the future either.

Good points here, and thanks - I would like to see this if you're up for doing it.

**Comment by gnieuwenhuis** [comment]
I've moved that function over, and also allowed setting some of the hyperparameters that were previously hard coded - to me it seemed like a good addition for anyone experimenting with the model in the future, but if you prefer not to expose those let me know.

**Comment by StephenChan** [comment]
Great idea, the parameters are certainly nice to have. Thanks!

Most of the tests look good, but `test_custom_batch_size()` might be missing the mark - the length of `ref_acc` verifies the number of epochs, but not the number of batches. Those two distinct numbers just coincidentally match in this test.

To assert on the number of batches, one idea is to use the `spy_decorator()` helper which is currently in `test_tasks.py`. The helper's probably best relocated to `tests/utils.py` instead of that file. Then I think something like this could be done:

```python
load_spied = spy_decorator(ImageLabels.load_data_in_batches)
with mock.patch.object(ImageLabels, 'load_data_in_batches', load_spied):
    ...    # Call trainer._train()
load_spied.mock_obj.assert_called_with(batch_size=50, random_seed=1)
```

Alternatively one could spy on `partial_fit()` to check its `call_count`. Either way is good IMO.

**Comment by gnieuwenhuis** [comment]
Nice catch! Check is fixed using the spy_decorator as you suggested, seems like a handy helper.

**Comment by StephenChan** [comment]
Great! Thanks. Lastly, I have a couple of comment updates/additions in mind. I might as well write that myself and commit it.

**Comment by StephenChan** [comment]
@gnieuwenhuis Anything else in mind before merging this?

After merging this, I can also release a new pyspacer version if that'll be useful soon.

**Comment by gnieuwenhuis** [comment]
Nothin more from me - I'm happy if you are.

It would be useful if you can release a new version, I can pin it in my draft PR's on the classifier and start moving forward with those

**Comment by StephenChan** [comment]
Sounds good, I'll let you know when the release is made.

**Comment by StephenChan** [comment]
@gnieuwenhuis 0.14.0 release has been made.
